src/coordination/coordinator_data.cpp

```diff
@@ -0,0 +1,215 @@
+  if (main_instance == registered_instances_.end()) {
+    return std::nullopt;
+  }
+
+  return CoordinatorInstanceStatus{.instance_name = main_instance->InstanceName(),
+                                   .socket_address = main_instance->SocketAddress(),
+                                   .is_alive = main_instance->IsAlive()};
+};
+
+auto CoordinatorData::ShowReplicas() const -> std::vector<CoordinatorInstanceStatus> {
+  std::shared_lock<utils::RWLock> lock{coord_data_lock_};
+  std::vector<CoordinatorInstanceStatus> instances_status;
+
+  for (const auto &replica_instance : registered_instances_ | ranges::views::filter(&CoordinatorInstance::IsReplica)) {
+    instances_status.emplace_back(CoordinatorInstanceStatus{.instance_name = replica_instance.InstanceName(),
+                                                            .socket_address = replica_instance.SocketAddress(),
+                                                            .is_alive = replica_instance.IsAlive()});
+  }
+
+  return instances_status;
+}
+
+auto CoordinatorData::SetInstanceToMain(std::string instance_name) -> SetInstanceToMainCoordinatorStatus {
+  // TODO: (andi) test this
+  std::lock_guard<utils::RWLock> lock{coord_data_lock_};
+
+  // Find replica we already registered
+  auto registered_replica = std::find_if(
+      registered_instances_.begin(), registered_instances_.end(),
+      [instance_name](const CoordinatorInstance &instance) { return instance.InstanceName() == instance_name; });
+
+  // if replica not found...
+  if (registered_replica == registered_instances_.end()) {
+    spdlog::error("You didn't register instance with given name {}", instance_name);
+    return SetInstanceToMainCoordinatorStatus::NO_INSTANCE_WITH_NAME;
+  }
+
+  // Stop for now because we need to swap success and fail callbacks
+  registered_replica->client_.StopFrequentCheck();
+
+  std::vector<CoordinatorClientConfig::ReplicationClientInfo> repl_clients_info;
+  repl_clients_info.reserve(registered_instances_.size() - 1);
+  std::ranges::for_each(registered_instances_,
+                        [registered_replica, &repl_clients_info](const CoordinatorInstance &replica) {
+                          if (replica != *registered_replica) {
+                            repl_clients_info.emplace_back(replica.client_.ReplicationClientInfo());
+                          }
+                        });
+
+  // PROMOTE REPLICA TO MAIN
+  // THIS SHOULD FAIL HERE IF IT IS DOWN
+  if (auto result = registered_replica->client_.SendPromoteReplicaToMainRpc(std::move(repl_clients_info)); !result) {
+    registered_replica->client_.StartFrequentCheck();
+    return SetInstanceToMainCoordinatorStatus::COULD_NOT_PROMOTE_TO_MAIN;
+  }
+
+  registered_replica->replication_role_ = replication_coordination_glue::ReplicationRole::MAIN;
+  registered_replica->client_.SetSuccCallback(main_succ_cb_);
+  registered_replica->client_.SetFailCallback(main_fail_cb_);
```

Comment:
Seems dangerous to start frequent check and then set the callbacks.
Why not set then and then start?

Reply:
Yes, that's wrong

Reply:
We didn't start it at all, that part is missing

Reply:
```
registered_replica->replication_role_ = replication_coordination_glue::ReplicationRole::MAIN;
  registered_replica->client_.SetSuccCallback(main_succ_cb_);
  registered_replica->client_.SetFailCallback(main_fail_cb_);
 // this is missing
  registered_replica->client_.StartFrequentCheck();
```